### PR TITLE
fix(data-table): fix sticky header issues

### LINKED ISSUE #6449: DataTable with stickyHeader styling of rows and headers regression
## What package(s) are you using?

- [x] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.
- Using a DataTable with the stickyHeader prop, there seems to have been a regression between 10.13 and 10.14, the position of header text and row text is no longer vertically centered in the row. The issue is visible in the storybook by toggling the stickyHeader knob - https://react.carbondesignsystem.com/?path=/story/datatable--default

- with stickyHeader:

    image

- Without stickyHeader:

    image

The styling with stickyHeader worked as expected in 10.13 / 7.13. CodeSandbox here: https://codesandbox.io/s/carbondatatablestickyheader-03096 , can change the dependency version in the sandbox to 10.15 / 7.15 to see the broken behaviour there as well.

> Is this issue related to a specific component?
- DataTable with stickyHeader prop

> What did you expect to happen? What happened instead? What would you like to see changed?
- Expected styling of header and row cells to be consistent with and without stickyHeader prop set. Would like styling to be restored so there is consistency.

> What browser are you working in?
- Chrome, firefox

> What version of the Carbon Design System are you using?
- issue in 10.14, 10.15 // 7.14, 7.15

> What offering/product do you work on? Any pressing ship or release dates we should be aware of?
- IBM Streams on CPD, a few months of leeway.

## Steps to reproduce the issue

1. Use DataTable with stickyHeader prop.

> Please create a reduced test case in CodeSandbox
- working with 10.13/7.13: https://codesandbox.io/s/carbondatatablestickyheader-03096 , switch dependency versions to 10.15/7.15 to see broken behaviour.

### PR DESCRIPTION
**Pull request by tw15egan**
Closes https://github.com/carbon-design-system/carbon/issues/6449

Fixes a spacing issue when `sticky header` is enabled in `DataTable`

#### Changelog

**Changed**

- Set specific padding for normal sticky header `td`, `th`
- Ensured other size variants padding is not affected

#### Testing / Reviewing

Ensure all variations of sticky header sizes look correct

### CODE REVIEW

File: packages/components/src/components/data-table/_data-table-core.scss [hunk 1]
```diff
@@ -520,6 +520,12 @@
       height: auto;
     }
 
+    &:not(.bx--data-table--compact) td,
```
Comment (joshblack) [inline packages/components/src/components/data-table/_data-table-core.scss]: Random, wasn't sure about this but would this be equivalent to:

```scss
&:not(.bx--data-table--compact, .bx--data-table--short, .bx--data-table--tall) td {
  // ...
}
```

?  Honestly wasn't sure how `not()` works lol

File: packages/components/src/components/data-table/_data-table-core.scss [hunk 2]
```diff
@@ -520,6 +520,12 @@
...
+    &:not(.bx--data-table--short) td,
```
Comment (joshblack) [inline packages/components/src/components/data-table/_data-table-core.scss]: Should we use `$prefix` in these selectors?

**Review by joshblack — CHANGES_REQUESTED** [review]

File: packages/components/src/components/data-table/_data-table-core.scss (hunk 1 above)
Comment (tw15egan) [reply packages/components/src/components/data-table/_data-table-core.scss]: Doesn't look like it's supported yet, but in the works 👍

**Comment by tw15egan** [comment]
@joshblack fixed 👍

**Comment by guychris** [comment]
@tw15egan Hi TJ, I had made a comment in the issue after taking a look at the storybook for this PR, this fixes normal tables but tables with expandable rows will need some work:

https://github.com/carbon-design-system/carbon/issues/6449#issuecomment-657768698

**Comment by tw15egan** [comment]
@guychris should be fixed now, thanks for catching that 👍

**Comment by guychris** [comment]
Thank you for the quick turnaround, just poking around some more and a few more edge cases I found:

- With expandable rows + stickyHeader, it looks like there's a bit of extra space at the bottom of the table, 
image
https://deploy-preview-6452--carbon-components-react.netlify.app/?path=/story/datatable--with-expansion



- With selection + sticky Header, the checkbox in the header row is not centered:
image
https://deploy-preview-6452--carbon-components-react.netlify.app/?path=/story/datatable--with-selection


- with overflow menu + sticky header, the overflow menu icon is not centered and it appears the alignment of the header columns is off:
image
https://deploy-preview-6452--carbon-components-react.netlify.app/?path=/story/datatable--with-overflow-menu


- one combination that I'm not sure how to test with your PR changes would be radio button selection + expandable rows + stickyHeader. Code for that in this sandbox, https://codesandbox.io/s/carbondatatablestickyheader-03096?file=/src/App.js

**Comment by tw15egan** [comment]
@guychris Thanks for taking the time and making sure I'm getting all the variants! 🙏 

Just pushed up some more changes that should take care of the rest of the issues. 

In regards to your last one, I checked how it renders with my changes, and it all looks good except the Table Headers are slightly misaligned with the Table Cells. Since there are 2 more cells than headers, you'd just have to set a manual offset on the headers based on the number of controls (in this case 2, so it'd probably be `48 * 2 px`. I didn't add this into the styles since I think this is an edge case that would be better handled at the application level. 


Let me know if I missed any more! 


@joshblack @andreancardona : added a conditional render in `TableHeader.js` so that we don't render a `bx--table-header-label` without any text. Instead, it will just render an empty `td`, which I used to target some very specific nodes: https://github.com/carbon-design-system/carbon/pull/6452/files#diff-977095845d8e1ee345eea1f4b7a52d9dR577

**Comment by guychris** [comment]
Thanks for the quick turnaround on the fixes @tw15egan , all the scenarios I was going through in the storybook that we might be using look good with stickyHeader now. It does look like changing the row height breaks a few different stylings unfortunately (apologies for opening the stickyHeader can of worms!). We're not changing row height in our case, but playing around in the storybook it's looking like the row height stylings are not being applied consistently to row elements.

few different parts I see that are not styled correctly:
- vertical alignment of header checkbox with tall rows is lower than header text
- compact row height without stickyHeader: 24px, with stickyHeader: 48px.
- short row height without stickyHeader: 32px, with stickyHeader: 51px.
- tall row heights without stickyHeader: 64px, with stickyHeader: 54px. 
- compact and short row height, row text and checkboxes are vertically aligned to the top instead of centered
- compact and short row height, overflow menu icon vertical alignment is at the top and the row separator line is out of alignment with the rest of the row.

some screenshots:
compact row heights with stickyHeader and overflow menu:
image

tall row heights with selection, header checkbox alignment:
image

**Comment by tw15egan** [comment]
@guychris just pushed up some more fixes. Keep in mind, that compact / short will break out of the specified size if the screen is too small to accommodate the full width.

**Review by joshblack — APPROVED** [review]

**Review by andreancardona — APPROVED** [review]

**Review by laurenmrice — CHANGES_REQUESTED** [review]
The column header type subtly shifts downward when toggling between enabled and the sticky header prop. I think it should be staying in the same place.
stickyheader

----------------------------------------------------------------------------------------------------------------------------

For the scrollbar on the right side, should the scrollbar extend to the column header area? Seems like it should only be present in the area that actually scrolls.
scroll

----------------------------------------------------------------------------------------------------------------------------

The scrollbar in Safari does not show up in full, and is cutoff at the top and bottom of the data table
cut

**Review by andreancardona — APPROVED** [review]

**Review by andreancardona — CHANGES_REQUESTED** [review]
@tw15egan I'm not entirely sure why but I am seeing a scroll bar in the header on Chrome

Screen Shot 2020-08-03 at 1 57 58 PM

**Comment by tw15egan** [comment]
@laurenmrice fixed the 1px issue. For the first screenshot, the whole area being scrolled is where the scroll bar is, the header is just fixed at the top. Technically the content is still there, just hidden under the header. The scrollbar positioning is all handled by the browser and is pretty tough to change, but I'll take a look at why Safari is not working properly 

@andreancardona fixed that issue 👍

**Comment by tw15egan** [comment]
@laurenmrice not sure what's going on in Safari, might be something with the deploy preview? This is what I see locally

### Mouse 

mouse

### Trackpad

trackpad

**Review by laurenmrice — CHANGES_REQUESTED** [review]
For the expanded data table stories, empty extra rows get added to the data table when the sticky header is activated.
Screen Shot 2020-08-03 at 2 12 44 PM

**Comment by tw15egan** [comment]
@laurenmrice should be fixed

**Review by andreancardona — APPROVED** [review]

**Review by laurenmrice — CHANGES_REQUESTED** [review]
The expansion stories looks good for the normal height data table now, but is getting height issues and icon misalignment for the other sizes. After that we should be good
Screen Shot 2020-08-04 at 10 26 03 AM

**Comment by tw15egan** [comment]
@laurenmrice hopefully all set now

**Review by laurenmrice — CHANGES_REQUESTED** [review]
Ok, I think this will be the last thing 😁!

For these stories `with dynamic content` , `with expansion` , `with options` : 
-The column header text is misaligned with the cell row text below it (for all sizes). 
-The stories mentioned with a column header checkbox icon get misaligned with the icons in the cell rows (for all sizes).

Screen Shot 2020-08-04 at 3 39 47 PM

**Review by laurenmrice — CHANGES_REQUESTED** [review]
`with dynamic content` and `with options` stories:
Upon expanding a row the checkbox in that row jumps lower for the short and compact sizes.

checkbox

**Review by laurenmrice — APPROVED** [review]
Woohoo! looks good to me 🙌🏻🎉

**Pull request merged**